### ISSUE TITLE
Mist memory model for Zag

### ISSUE THREAD

**Issue by Shaping** [dvmason/Zag-Smalltalk#2]
Re: https://www.youtube.com/watch?v=iw4FPqe8KdY&t=1093s 22:05 to 36:05

The above section of Martin McClure's Mist video describes a very simple, fast, exponential memory model involving binary-powers-of-2-sized bins for allocating objects.  We need not lift this GC mechanism into Zag itself, at this early stage, as he is proposing for Mist.  We could instead implement this GC design as C primitives on the VM side.  In this way, objects on the heap will never move.  This fact alone will increase execution speed and create more deterministic latencies.  

Thoughts anyone?

**Comment by dvmason** [comment]
Thanks for your thoughts! This is very much a work in progress, but the documentation has some pieces on this (although not well organized).
The current plan is:
1) each execution thread has its own generational, copying collector with a couple of arenas
2) there is a global space which is non-moving, and organized as a Fibonacci allocation - similar to Martin's binary allocator (but with less internal fragmentation) - there is a separate collector thread which manages this, interacting with the execution threads to find roots
I did like Martin's idea of having large objects mmap/munmap-ed - I need to think on this a while.

**Comment by Shaping** [comment]
(I just realized that I should have said "Zig" instead of "C" primitives" above.)

Yes, agreed, regarding handling of large objects via mmap.  He also seemed to say that the binary allocation strategy may need to be replaced by reference counting.  I'm not sure I follow that.  He seems to be saying that merely flipping the marked/unmarked bit for each fixed-size block of bytes in each bin won't scale. Not sure.  It looks pretty fast even if the bins are large.

I've read through most all of the online doc for Zag.   I must have missed the part about memory management.

I'm waiting now 27 years for a performant Smalltalk.  It may actually happen.

Are doing this solo?  Are your grad students helping?

Unison (https://www.unison-lang.org/) looks like the future of programming language, at least in terms of how the code is stored and versioned (AST is hashed; no source is stored).  Function names are attached to hashes, and renaming is easy.  Nothing breaks via renaming.  The codebase always works, and experiments/forks are very.   Zag is already very AST-centric. 
We could steer Zag in this direction.  Have you considered this?  

I like Unison so much that I'm willing to tolerate the much less interesting (non-Smalltalk, non-anglogrammatical, non-keyword syntax) long enough perhaps to change the Unison parser (written in Haskell) to "support" or at least fake-out a message-sending like syntax.  I prefer a functional/declarative domain-level grammar.  So I can either morpho Unison to be more like Smalltalk, or morph Smalltalk to use hashing broadly.  Zag looks like a great place to do that.   I'm trying to get rid of the VM (as Martin is suggesting) and make memory management very simple and fast (I don't mind some wasted memory).  

Can I currently play with a rough version of Zag working on some version of Pharo?  Are you hosting Zag development in Pharo , or just trying to parallel Pharo class structure? 

I'm currently stuck in VW 7.10/8.3.  Zak seems like a good reason to port to Pharo.

**Comment by dvmason** [comment]
Unfortunately, I am doing this mostly solo at the moment. I have a grad student starting in September, but it will be a while before they are up to speed.

See https://github.com/dvmason/Zag-Smalltalk/blob/main/Documentation/Garbage_Collection.md - there's a few typos/unclarities, but the gist is there.

Unison does look interesting. I wish them well.

I'm supposed to give a talk at the UK Smalltalk Meetup (hopefully in September). I'm hoping to be able to export code to run on Zag by then.

**Comment by dvmason** [comment]
Also, note that Mist is *very* single-threaded... there aren't even any processes.

Whereas Zag is intended to be very multi-threaded... using OS threads, and it will do blocking I/O in I/O threads.

**Comment by Shaping** [comment]
Yes, we want parallelizing of algorithms, which is very easy to do in a purely functional language (Unison, namely).  Unison easily parallelizes across cores/OS-processes and machine nodes.  The big idea here is that is can economically move the code (always via hashed AST) to the data sitting on another node, instead of moving potentially very large data around to where the code runs.  This code syncing need happen only once; the code then cashed on the other node(s).

They've designed the so called Unison abilities to handle algebraic effects.  The separation of the functional chain and effects (side-effects) is clean, but the syntax could be somewhat more approachable.   I don't want to trade accurate, low-effort readability for the transparent, easy parallelizing.  We should be able to do some similar at least as well in Zag.

So Zag will be multiprocess/multicore and also able to run multiple threads in one OS process?  I hope the Zag syntax for parallelizing stays simple, and just makes it work in the most efficient way possible.

**Comment by Shaping** [comment]
Why do most  GC designs by default include operations to copy and move?  The binary-power-sized bins strategy involves no copies or moves.  The memory blocks in each bin are quickly accessed via a linked list.   A bit is flipped to indicate allocated for use by an object or not.

If wasted memory is a concern, recursively apply the  binary-power-sized bin structure to the leftover portion of each memory block, taking care to collect for the secondary (or n-ary) bin linked list only those blocks whose leftovers are similar in size by some predefined percentage.  One or two nestings of this pattern will will leave very little bin-block memory unused/inaccessible.   I don't see a good reason for the current complexity and slowness of popular GC algos.    Someone please explain how the binary-power-sized bins strategy fails to be fast and scalable.

**Comment by Shaping** [comment]
Yes, the Fib size progression makes more sense than geometric.

Would like to see more detail on this and on how the arenas coordinate.

**Comment by Shaping** [comment]
Dave, do you need any help with the Zig code?  Are you satisfied with Zig's simplicity and superiority relative to regular crap C?

Have you timed the GC copies you mentioned?  In general, I want the newest and best VM for Smalltalk (or any message-based language) to be temporally introspective, by design, easing measurement of run-times for any chunk of code/block or actor network performing a task.  Are you building such an ability it to the VM, or at least considering it?  I'm trying to include time measurements in all my SUnit tests, which is an unreliable exercise in the best cases, given how most GCs work.  The measured times vary too much.  This is why a spread of  Fib-sized bins of known sizes is interesting.   As much as we all want fast allocating and freeing, I also to see more deterministic measured run-times and smaller variances.  This helps when we program  and test  microcontrollers/IoT devices.

**Comment by Shaping** [comment]
Is Zag intended to be compatible with the exiting Pharo image format?

**Comment by dvmason** [comment]
I've been updating some of the documentation, so looking at recent changes may be helpful. I've particularly updated some of the MemoryManagement, and have integrated Mist-like large-object allocation, so that now anything larger than about 4K words (32KiB) gets allocated its own pages for the content (great for reading in a whole file - it can simply be mmapped)

Zag Smalltalk is also getting extended to Zag Dynamic Runtime, as I've realized that I can easily support other dynamically-typed languages like Ruby and Python easily, but also Javascript, Self, and APL with a bit of work, and Scheme with a bit more work. This would support a lot more experimentation with other ideas about dynamic programming languages. It will provide: 
 - immediate 64-bit tagged values supporting 51-bit integers, double-precision floating point, symbols, Unicode characters, booleans, nil/null with room for expansion;
 - full OS-thread support to exploit multi-code/many-core architectures;
 - a production-quality parallel/concurrent garbage collector and high-performance memory allocation;
 - efficient support for the "challenging bits" of the languages above including full closures, call-with-current-continuation, become:, allInstances, n-dimensional arrays;
 - efficient virtual method dispatch;
 - threaded, fully live-debuggable, execution model;
 - multi-language debugging (at least within a paradigm);
 - type-interpretation;
 - JIT code generation, probably based on LLVM;
 - exporting of executable programs to a range of platforms;
 - leveraging GPU execution engines.
The goal is that a new language implementation would require only a parser and possibly tweaks to the code generation to capture idiosyncratic execution models. Much of the design for this model is done, and implementation has started on the basics. The full vision will take several years.

As for help, I'd love help! Right now I need to get a few more details nailed down, but implementation of primitives is a large, but highly granular task.

No, Zag will not he compatible with the existing Pharo image format, but Pharo is now built from pieces and it might be possible to build a Zag image from many of the same pieces, plus a few Zag-specific ones. Alternately a converter could take a OpenSmalltalk image and produce a Zag image.

**Comment by Shaping** [comment]
It's a splendid feature list.  I'm re-reading the docs.

Are we committed to being temporally accountable?  There will be no robot-control or IoT without it.  Can we build accurate run-time measurement into the VM, so that it's easy and desirable use, without clutter the code?   Perhaps have a global switch for turning it on and off.
